services/statestorage/adapter/persistence.go

```diff
@@ -15,7 +15,7 @@ type ContractState map[string]*protocol.StateRecord
 type ChainState map[primitives.ContractName]ContractState
 
 type StatePersistence interface {
-	Write(height primitives.BlockHeight, ts primitives.TimestampNano, root primitives.Sha256, diff ChainState) error
+	Write(height primitives.BlockHeight, ts primitives.TimestampNano, proposer primitives.NodeAddress, root primitives.Sha256, diff ChainState) error
 	Read(contract primitives.ContractName, key string) (*protocol.StateRecord, bool, error)
-	ReadMetadata() (primitives.BlockHeight, primitives.TimestampNano, primitives.Sha256, error)
+	ReadMetadata() (primitives.BlockHeight, primitives.TimestampNano, primitives.NodeAddress, primitives.Sha256, error)
```

Comment:
if we end up keeping the proposer here, I'd like you to introduce a struct for the metadata